src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java

```diff
@@ -68,6 +68,9 @@ public class StatementRestrictions
 
     public static final String INDEX_DOES_NOT_SUPPORT_ANALYZER_MATCHES_MESSAGE = "Index on column %s does not support ':' restrictions.";
 
+    public static final String ANN_DOES_NOT_SUPPORT_FILTERING_MESSAGE = "ANN does not support filtering.";
```

Comment:
IIUC it is OK if there is additional index-based filtering (@jbellis please correct me if I'm wrong).
If so, then we should say sth along "ANN requires index on filtered columns; please create an index using CREATE INDEX ..." rather than "ANN doesn't support filtering"

Reply:
Yes.  (More specifically, it requires an SAI index.)

Reply:
Is it okay if there is filtering on clustering columns?

Reply:
My earlier comment was in error.  The intent is

- You can't ORDER BY ... ANN OF without an index
- vector indexes are only for ANN, not for equality predicates

I'm okay with allowing ALLOW FILTERING queries against vector columns but only if it doesn't make our lives harder elsewhere (if we have to fix anything in SelectStatement, it's not worth it).

Reply:
I quickly looked at the code and the code look like it should not work as the Operator logic for the ANALYZER should fire an exception. After it might not be the latest version of the code.

Reply:
> You can't ORDER BY ... ANN OF without an index

This is already covered [here](https://github.com/datastax/cassandra/blob/a9afb78a82655d7acde1d5b9e21f4dbcf4265799/src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java#L600-L609).

> vector indexes are only for ANN, not for equality predicates

This is handled correctly [here](https://github.com/datastax/cassandra/blob/a64160f506609b20d8cd34609fe6b7ca255dddf6/src/java/org/apache/cassandra/index/sai/IndexContext.java#L418-L422).

In general, this PR is meant to prevent an edge case where ALLOW FILTERING would otherwise be necessary and would give correct results, but won't (for now) because of how ANN search works. Here is a sample set of queries that show the issue:

```
token@cqlsh> select * from test.tabe1 ;

 p | c | v   | vec
---+---+-----+------------
 1 | 2 | 123 | [0.3, 0.3]
 0 | 1 |  17 | [0.5, 0.5]

(2 rows)

 

token@cqlsh> select * from test.tabe1 ORDER BY vec ANN OF [1.0, 2.0] LIMIT 10;

 p | c | v   | vec
---+---+-----+------------
 1 | 2 | 123 | [0.3, 0.3]
 0 | 1 |  17 | [0.5, 0.5]

(2 rows)

token@cqlsh> select * from test.tabe1 WHERE v > '1' ALLOW FILTERING ; 

 p | c | v   | vec
---+---+-----+------------
 1 | 2 | 123 | [0.3, 0.3]
 0 | 1 |  17 | [0.5, 0.5]

(2 rows)

 

token@cqlsh> select * from test.tabe1 WHERE v > '1' ORDER BY vec ANN OF [1.0, 2.0] LIMIT 10 ALLOW FILTERING;

 p | c | v | vec
---+---+---+-----

(0 rows)
```

I reproduce the issue when `v` is a non clustering column, and it appears to work when it is a clustering column. I just added a commit with a test showing a case where ANN and ALLOW FILTERING (appear) to work.

I am working on a more nuanced rejection now.

Reply:
oh, right.  that's because filtering only happens after index predicates, and ANN is executed as part of index predicates

I *thought* the clustering columns were also applied after index predicates, so I'm not sure why that works.

Spend a morning tracking that down if you want but fixing this is not a priority, I'm fine with the original plan of "just disallow filtering with ANN and move on"

Reply:
Thinking about it, it's probably more subtle than just order of operations, but again, I'm fine with just blanket-disallowing it

Reply:
It does appear to be pretty subtle because while clustering columns seem to work, partitioning columns do not. The following test fails on the last line because the query returns no rows.

```java
    @Test
    public void testAllowFilteringWithMediumComplexityPrimaryKey() throws Throwable
    {
        createTable("CREATE TABLE %s (pk int, b int, v vector<float, 3>, PRIMARY KEY(pk, b))");
        createIndex("CREATE CUSTOM INDEX ON %s(v) USING 'StorageAttachedIndex'");
        waitForIndexQueryable();

        execute("INSERT INTO %s (pk, b, v) VALUES (0, 0, [1.0, 2.0, 3.0])");
        execute("INSERT INTO %s (pk, b, v) VALUES (1, 2, [2.0, 3.0, 4.0])");
        execute("INSERT INTO %s (pk, b, v) VALUES (2, 4, [3.0, 4.0, 5.0])");
        execute("INSERT INTO %s (pk, b, v) VALUES (3, 6, [4.0, 5.0, 6.0])");

        // Note that this type of search only appears to work when b is a clustering column
        assertRows(execute("SELECT pk FROM %s WHERE b > 2 ORDER BY v ANN OF [2.5, 3.5, 4.5] LIMIT 10 ALLOW FILTERING;"),
                   row(2), row(3));

        assertRows(execute("SELECT pk FROM %s WHERE pk > 2 ORDER BY v ANN OF [2.5, 3.5, 4.5] LIMIT 10 ALLOW FILTERING;"),
                   row(3));
    }
```

I'll spend a little more time looking through this to see if I can get a grasp on the underlying issues. I want to make more progress on vsearch 14, so it might be worth a generic rejection that we can come back to later and improve when it is a priority to accept queries with ALLOW FILTERING.

Reply:
(Might be a permanently low priority, ALLOW FILTERING is mostly a hack to allow exploring tiny datasets from before we had useful indexes)

Reply:
> I thought the clustering columns were also applied after index predicates, so I'm not sure why that works.

Good point. It was only working because the limit on ANN was large enough to get all rows. The following query returns no rows when run against the test in my last comment. This suggests we should prevent these queries too since we cannot guarantee correctness.

```cql
SELECT pk FROM %s WHERE b > 2 ORDER BY v ANN OF [1,2,3] LIMIT 2 ALLOW FILTERING;
```

The one detail I don't understand yet is why there is a difference between partition and cluster columns. In setting up a break point at the getIndexes method, it appears that the partition restriction becomes an expression even though it doesn't have an index. Looking at that now.

Reply:
Partition does get applied as part of index processing, although I'm not sure where the code to make that happen lives.

Reply:
The difference with partition key and clustering column restrictions is due to the difference in these two code blocks:

https://github.com/datastax/cassandra/blob/0a7cb8aee34a61c1b75def1b7ba46be9a3703e8f/src/java/org/apache/cassandra/cql3/restrictions/ClusteringColumnRestrictions.java#L130-L150

https://github.com/datastax/cassandra/blob/0a7cb8aee34a61c1b75def1b7ba46be9a3703e8f/src/java/org/apache/cassandra/cql3/restrictions/PartitionKeySingleRestrictionSet.java#L125-L136

The clustering column reference does not add the restriction to the row filter, but the partition key does. As a result, the partition key restriction gets added to the SAI query, which is wrong but benign in all query examples that I can think of _except_ vector due to the way we search. In the vector case, we try to find rows matching the partition key restriction, but there are not any because the view is empty (the partition key is not indexed).

So, from a correctness perspective, this is not a bug, but I assume it's still something to fix.

Either way, neither query works correctly, so we'll block it.

Reply:
It looks like this might be even more complex. We also add a `ClusteringIndexFilter` as a data range. I think this might be why it "works" in the case where we get all of the relevant rows. In general, this code seems fragile, but does appear to work. The problems materialize when we want to apply ORDER and LIMIT based on indexed columns.

I'll refactor this PR to reject any query that has restrictions on columns without an index.